### PR TITLE
[feat]: Log number of broken records

### LINKED ISSUE #1034: [Feature request] Log number of datasets records on which fetch failed
Sometimes the fetch process fails (for example, like in this issue, and there are no way to observe it except for reading through logs.  Sometimes it is not even a Speculators bug, like in this [issue](https://github.com/vllm-project/vllm/issues/44485). 

I would like to see number of records among other graphs on Tensorboard or similar experiment tracking system.

### PR DESCRIPTION
**Pull request by pavelgein**
## Purpose
Log number of broken records. Closes  #1034

## Tests
I have run and see scalars at the experiment tracking system

## Checklist

I have filled in:

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [ ] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.

### CODE REVIEW

File: src/speculators/train/trainer.py [hunk 1]
```diff
@@ -469,6 +469,9 @@ def train_epoch(self, epoch: int):
             loss.backward()
             torch.nn.utils.clip_grad_norm_(self.model.parameters(), 1.0)
 
+            metrics["error_records_sum"] = torch.tensor(batch["error_records"], dtype=torch.int32, device=loss.device)
+            metrics["error_records_total"] = torch.tensor(1.0 if (self.is_distributed and self.rank == 0) else 0, device=loss.device)
```
Comment (fynnsu) [inline src/speculators/train/trainer.py]: I think we want
`torch.tensor(1.0 if self.rank == 0 else 0, device=loss.device)`

Otherwise the non-distributed case might have a division by zero.

**Review by fynnsu — APPROVED** [review]
Approved but please correct issue mentioned below.

File: src/speculators/train/trainer.py (hunk 1 above)
Comment (pavelgein) [reply src/speculators/train/trainer.py]: Thank you for the review, fixed it and style

File: src/speculators/train/data.py [hunk 2]
```diff
@@ -390,4 +405,6 @@ def __call__(self, batch: Sequence[BatchType | None]) -> BatchType:
         # shape: [1, max_len]
         collated_data["document_ids"] = document_ids
 
+        collated_data["error_records"] = num_dropped
```
Comment (fynnsu) [inline src/speculators/train/data.py:408]: Looks like we actually need this to be a tensor at this stage. (Everything coming out of the collation function should be a tensor).

**Review by fynnsu — COMMENTED** [review]
The tests are failing, please see below.

File: src/speculators/train/data.py (hunk 2 above)
Comment (pavelgein) [reply src/speculators/train/data.py:408]: Well, it's interesting, because it successfully running. And the [code](https://github.com/vllm-project/speculators/blob/main/src/speculators/train/trainer.py#L546-L551) is written in the way to handle non-tensor items in the batch.

It seems that the problem is with the test `tests/unit/train/test_data.py::test_collate_fn_basic` itself, because it expects that all items in a batch are tensors. 
So I suggest that I fix the test.

Comment (fynnsu) [reply src/speculators/train/data.py:408]: Yes, however it's not just the one unit test that's failing. There's also a bunch of integration tests https://buildkite.com/vllm/speculators/builds/1315/list?jid=01a03e69-428c-4c12-ab93-b4c7342a0e9e&tab=output.

These mostly all use the same batch setup logic, so you can certainly fix it there. Either way works.

Comment (pavelgein) [reply src/speculators/train/data.py:408]: Yes, this integration tests lacks filtering logic from Trainer class

Comment (pavelgein) [reply src/speculators/train/data.py:408]: Fixed in [ce604b7](https://github.com/vllm-project/speculators/pull/1035/commits/ce604b7dda2f3bfdac3f9cd39c707217388acae6)

File: tests/integration/conftest.py [hunk 3]
```diff
@@ -337,7 +337,8 @@ def make_batch(
         preprocess=preprocess,
     )
     batch = collate_fn(samples)
-    return {k: v.to(device) for k, v in batch.items()}
+    # We drop all non-tensors items, to mimic logic from Trainer class, see gpu_batch formation
+    return {k: v.to(device) for k, v in batch.items() if isinstance(v, torch.Tensor)}
```
Comment (fynnsu) [inline tests/integration/conftest.py]: The Trainer class doesn't drop the keys, they just only move them `.to(device)` if they're tensors:
```
gpu_batch = {
        k: v.to(self.local_rank, non_blocking=True)
        if isinstance(v, torch.Tensor)
        else v
        for k, v in batch.items()
    }
```

Comment (pavelgein) [reply tests/integration/conftest.py]: Do you mean, that comment is misleading or fix should be modified?

Comment (fynnsu) [reply tests/integration/conftest.py]: I think the fix should be modified. This line is currently filtering out non-tensors, but we should make the logic match the trainer class logic and just move the tensors to device while keeping the non-tensors.

Comment (pavelgein) [reply tests/integration/conftest.py]: But trainer class drops the keys in the sense that they will not be passed to model's forward, since not all keys from `batch` will be in `gpu_batch` (and gpu_batch will be passed to model's forward). 

As a solution, I can patch all tests that uses make_batch that they will pass only tensors to model's forward

Comment (pavelgein) [reply tests/integration/conftest.py]: As far as I can see, all the uses of make_batch directly passed to model's forward.
IMHO, if this "dropping" logic will be copied to all tests (well, there are not many of them), it will just decrease readability. But you're the maintainer, I will adjust code

Comment (fynnsu) [reply tests/integration/conftest.py]: Sorry, maybe I'm misunderstanding something but I'm pretty sure all the keys in `batch` end up in `gpu_batch` from this logic: https://github.com/vllm-project/speculators/blob/331a5f887f3a70bade6129fb7a5865f498f37d45/src/speculators/train/trainer.py#L546-L551

The way I understand this is if we have a batch with two kv pairs `{"key1": tensor([5]), "key2": 0}` where one is a tensor and the other is a non-tensor (in this case an `int`), we'll get a `gpu_batch` of `{"key1": tensor([5], device="cuda:0"), "key2": 0}`. i.e. the tensor will be copied to the gpu, and the non-tensor values will be left as is. And then `gpu_batch` gets passed directly to the model function.

In the version of the dict comprehension you have in conftest you'll instead get `{"key1": tensor([5])}`, with `key2` filtered out.

I just want the test batch creation logic to match the trainer logic, so that we're always keeping all of the keys.

Comment (pavelgein) [reply tests/integration/conftest.py]: Yeah, you are right and I am not, I don't know how I missed the else clause several times in a row.

Comment (pavelgein) [reply tests/integration/conftest.py]: Fixed in 85e38c9, thanks again

**Review by fynnsu — APPROVED** [review]
Thanks!
